contracts/ERC721Collateralizer.sol

```diff
@@ -0,0 +1,372 @@
+        require(collateralTokenAddress != address(0));
+
+        /*
+            Ensure that the agreement has not already been collateralized.
+
+            If the agreement has already been collateralized, this check will fail
+            because any valid collateralization must have some sort of valid
+            address associated with it as a collateralizer.  Given that it is impossible
+            to send transactions from address 0x0, this check will only fail
+            when the agreement is already collateralized.
+        */
+        require(agreementToCollateralizer[agreementId] == address(0));
+
+        ERC721 erc721token = ERC721(collateralTokenAddress);
+        address custodian = address(this);
+
+        // Ensure that the collateralizer is the owner of this token ID.
+        require(erc721token.ownerOf(collateralTokenID) == collateralizer);
+
+        // The token must be approved for transfer by this contract.
+        require(erc721token.getApproved(collateralTokenID) == custodian);
+
+        // Store collaterallizer in mapping, effectively demarcating that the
+        // agreement is now collateralized.
+        agreementToCollateralizer[agreementId] = collateralizer;
+
+        // Transfer the collateral asset to this contract.
+        erc721token.transferFrom(collateralizer, custodian, collateralTokenID);
+
+        // Emit an event that collateral has been secured.
+        CollateralLocked(agreementId, collateralTokenAddress, collateralTokenID);
+
+        return true;
+    }
+
+    /**
+     * Returns collateral to the debt agreement's original collateralizer
+     * if and only if the debt agreement's term has lapsed and
+     * the total expected repayment value has been repaid.
+     *
+     * @param agreementId bytes32 The debt agreement's ID
+     */
+    function returnCollateral(
+        bytes32 agreementId
+    )
+        public
+        whenNotPaused
+    {
+        // The address of the ERC721 contract.
+        address collateralTokenAddress;
+        // The token ID, associating the collateral token and the given ERC721 contract.
+        uint256 collateralTokenID;
+        // The terms contract according to which this asset is being collateralized.
+        TermsContract termsContract;
+
+        // Fetch all relevant collateralization parameters.
+        (
+            collateralTokenAddress,
+            collateralTokenID,
+            termsContract
+        ) = retrieveCollateralParameters(agreementId);
+
+        // Ensure a valid form of collateral is tied to this agreement id.
+        require(collateralTokenAddress != address(0));
+
+        // Withdrawal can only occur if the collateral has yet to be withdrawn.
+        // When we withdraw collateral, we reset the collateral agreement
+        // in a gas-efficient manner by resetting the address of the collateralizer to 0.
+        require(agreementToCollateralizer[agreementId] != address(0));
```

Comment:
So there's sort of a re-entrancy smell here -- it's *not* a viable attack right now for a whole bunch of reasons (i.e. mainly, the fact that the terms contract can only be collateralized if its permissioned), but as a matter of principle / extensibility / preventing anyone who is updating this code in the future from making a mistake, I think it should be addressed.

Generally, if a `require` statement asserts some sort of crucial boolean invariant (e.g. collateral has been withdrawn), and then, at some point later, that boolean is flipped, there should be *no* untrusted contract calls between the `require` statement and the boolean flipping.  In this case, we make a "untrusted" call to the terms contract before actually flipping the boolean.  This is fine, because the termsContract can't be untrusted here, but, for the reasons I spelled out before, I think this invariant should be `require`'d immediately before the `delete` statement.